tests/Serval.SpecFlowTests/StepDefinitions/MachineApiStepDefinitions.cs

```diff
@@ -0,0 +1,337 @@
+            new TranslationEngineConfig
+            {
+                Name = user,
+                SourceLanguage = source_language,
+                TargetLanguage = target_language,
+                Type = engineTypeString
+            }
+        );
+        EnginePerUser.Add(user, engine.Id);
+        ClearPretranslationConfig();
+    }
+
+    [When(@"a (.*) corpora containing (.*) are added to (.*)'s engine in (.*) and (.*)")]
+    public async Task GivenCorporaForEngineNoTranslate(
+        string fileFormatString,
+        string filesToAddString,
+        string user,
+        string sourceLanguage,
+        string targetLanguage
+    )
+    {
+        await GivenCorporaForEngine(
+            fileFormatString,
+            filesToAddString,
+            user,
+            sourceLanguage,
+            targetLanguage,
+            pretranslate: false
+        );
+    }
+
+    [When(@"a (.*) corpora containing (.*) are added to (.*)'s engine in (.*) to translate into (.*)")]
+    public async Task GivenCorporaForEngineToTranslate(
+        string fileFormatString,
+        string filesToAddString,
+        string user,
+        string sourceLanguage,
+        string targetLanguage
+    )
+    {
+        await GivenCorporaForEngine(
+            fileFormatString,
+            filesToAddString,
+            user,
+            sourceLanguage,
+            targetLanguage,
+            pretranslate: true
+        );
+    }
+
+    public async Task GivenCorporaForEngine(
+        string fileFormatString,
+        string filesToAddString,
+        string user,
+        string sourceLanguage,
+        string targetLanguage,
+        bool pretranslate
+    )
+    {
+        if (!Enum.TryParse(fileFormatString, ignoreCase: true, result: out FileFormat fileFormat))
+            throw new ArgumentException(
+                "Corpus format type needs to be one of: " + string.Join(", ", EnumToStringList<FileFormat>())
+            );
+        var filesToAdd = filesToAddString.Split(", ");
+        var engineId = await GetEngineFromUser(user);
+
+        var corpusId = await PostCorpusToEngine(
+            engineId,
+            fileFormat: fileFormat,
+            filesToAdd,
+            sourceLanguage,
+            targetLanguage,
+            pretranslate
+        );
+    }
+
+    [When(@"(.*)'s engine is built")]
+    public async Task WhenEngineIsBuild(string user)
+    {
+        var engineId = await GetEngineFromUser(user);
+        var newJob = await translationEnginesClient.StartBuildAsync(engineId, translationBuildConfig);
+        int cRevision = newJob.Revision;
+        while (true)
+        {
+            var result = await translationEnginesClient.GetBuildAsync(engineId, newJob.Id);
+            if (!(result.State == JobState.Active || result.State == JobState.Pending))
+            {
+                // build completed
+                break;
+            }
+            Thread.Sleep(500);
```

Comment:
What is the purpose of this sleep? Also, it would be better to use `await Task.Delay(500);`.

Reply:
Throttling requests.  Added comment.